### PR TITLE
Program nexthop table as a ternary match.

### PR DESCRIPTION
**Pull request by n-sandeep**
Earlier, nexthop table was exact match, due to LAG addition, nexthop table is moved to WCM block, ternary match. Changes in krnlmon is made to program nexthop table as ternary match

### CODE REVIEW

**Review by vsureshkumarp — APPROVED** [review]
Changes looks good to me.

**Review by nupuruttarwar — APPROVED** [review]
LGTM

**Review by aashishkuma — COMMENTED** [review]
This change is already part of the LAG PR.
https://github.com/ipdk-io/krnlmon/pull/54

**Review by ffoulkes — APPROVED** [review]
LGTM

**Review by ffoulkes — CHANGES_REQUESTED** [review]
Afterthought: the commit comment says, "nexthop table is moved to WCM block, ternary match." Does this mean that this change is dependent on a change in the P4 program?

If so, and assuming that `tdi_key_field_set_value_and_mask()` doesn't take care of this internally, we have a design issue: tight coupling between implementation details of a specific P4 program and the Kernel Monitor source code.

What will happen if there's a mismatch? Will the user receive a comprehensible error message? If so, how will they receive it?

**Comment by n-sandeep** [comment]
> tdi_key_field_set_value_and_mask

yes, this change is added in LAG PR as well. But we have an HSDES for LNW feature, because of overlay ping failure. 
Need to fix the issue seperately, as LAG PR is taking some more time to get merged.

**Comment by n-sandeep** [comment]
> Afterthought: the commit comment says, "nexthop table is moved to WCM block, ternary match." Does this mean that this change is dependent on a change in the P4 program?
> 
> If so, and assuming that `tdi_key_field_set_value_and_mask()` doesn't take care of this internally, we have a design issue: tight coupling between implementation details of a specific P4 program and the Kernel Monitor source code.
> 
> What will happen if there's a mismatch? Will the user receive a comprehensible error message? If so, how will they receive it?

Yes Derek, since there is a change in p4 program, where nexthop table has been moved to WCM block from SEM block (exact match).
TDI provides seperate API's for each block. Since Ternary/WCM needs a MASK to be populated we need to use a newer API which can take mask as a parameter. 
I believe TDI API cannot be same for both ternay and exact match, as exact match doesnt require any mask, but ternary needs a mask that is defined by user. In this case mask can be anything in the 16 bits defined.

If user tries to program a ternary match type with TDI API other than the one used in the PR, we will see an TDI error which comes back to krnlmon and entry doesnt get programmed.

**Comment by ffoulkes** [comment]
@n-sandeep
I don't have a problem with there being separate API calls. My concern is that:

1. Once this update is committed, P4CP will no longer work with older versions of the P4 Program.
2. In the event of a mismatch, the user won't get meaningful feedback that helps them understand what the problem is.

**Comment by n-sandeep** [comment]
> @n-sandeep I don't have a problem with there being separate API calls. My concern is that:
> 
> 1. Once this update is committed, P4CP will no longer work with older versions of the P4 Program.
> 2. In the event of a mismatch, the user won't get meaningful feedback that helps them understand what the problem is.

You are right Derek, with the current TDI API mechanism, we cannot have a backward compatability as the match type in p4 program and TDI API are tightly coupled. When match type changes, we need to modify/use new TDI API to program the entry.

Atleast P4CP doesn't have control over this, how do we achieve backward compatability ?

**Comment by ffoulkes** [comment]
> At least P4CP doesn't have control over this, how do we achieve backward compatibility ?

1. We add support for the new action, while retaining support for the old action. (Open Closed Principle)
2. We introduce a conditional that Kernel Monitor can use to choose between the two actions.
3. We provide a mechanism that can be used to specify which of the two actions to take.
4. We add a new conditional whenever a new decision needs to be made.

In a nutshell, _Kernel Monitor is configurable, just like the pipeline_.

The quick-and-dirty approach is to introduce a new C header file that #defines symbols specifying compile-time conditionals. You'd have to edit the file and rebuild krnlmon to apply the change, but at least it wouldn't be baked in.

Next step up would be introduce a configuration file that is read at startup time (or injected via gRPC) and used to set run-time conditionals. We could include an appropriate configuration file in the same directory as the P4 program.

Better still would be to use annotations in the P4 program to specify the appropriate metadata.

**Comment by n-sandeep** [comment]
> > At least P4CP doesn't have control over this, how do we achieve backward compatibility ?
> 
> 1. We add support for the new action, while retaining support for the old action. (Open Closed Principle)
> 2. We introduce a conditional that Kernel Monitor can use to choose between the two actions.
> 3. We provide a mechanism that can be used to specify which of the two actions to take.
> 4. We add a new conditional whenever a new decision needs to be made.
> 
> In a nutshell, _Kernel Monitor should be configurable, just like the pipeline_.
> 
> The quick-and-dirty approach is to introduce a new C header file that #defines symbols specifying compile-time conditionals. You'd have to edit the file and rebuild krnlmon to apply the change, but at least it wouldn't be baked in.
> 
> Next step up would be introduce a configuration file that is read at startup time (or injected via gRPC) and used to set run-time conditionals. We could include an appropriate configuration file in the same directory as the P4 program.
> 
> Better still would be to use annotations in the P4 program to specify the appropriate metadata.

Yes Derek, annotations will come as part of next releases. we can see if such backward compatability achieveed with annotations.

File: switchapi/es2k/switch_pd_routing.c
[the remark this answers is not in the document]
Comment (vsureshkumarp) [reply switchapi/es2k/switch_pd_routing.c:1]: Let me share my opinion here - 
A key prerequisite for the networking recipe involves utilizing the networking recipe program associated with a specific source SHA/version we provided.
The entire use case, as provided by us, is not anticipated to undergo any alterations by the user during their experimentation. 
I don't see a necessity to manage backward compatibility when the complete use-case specifications, including the P4 program, source code, and READMEs related to setting up and configuring the rules, are tailored to a particular version of the source code.

In fact, above macro based approaches creates a lot of confusion to the end user in my mind. 

Summary is I don't see any need to maintain a backward compatibility based on the current design.

**Review by ffoulkes — APPROVED** [review]
LGTM

See comment regarding the text to go in the commit comment and the text to go into the README file.

**Pull request merged**